lib/https.js

```diff
@@ -273,19 +275,24 @@ function request(options, cb) {
           'DeprecationWarning', 'DEP0109');
       }
     }
-  } else if (options && options[searchParamsSymbol] &&
-             options[searchParamsSymbol][searchParamsSymbol]) {
+  } else if (args[0] && args[0][searchParamsSymbol] &&
+             args[0][searchParamsSymbol][searchParamsSymbol]) {
     // url.URL instance
-    options = urlToOptions(options);
-  } else {
-    options = util._extend({}, options);
+    options = urlToOptions(args.shift());
+  }
+
+  if (args[0] && typeof args[0] !== 'function') {
+    options = util._extend(options, args.shift());
   }
+
   options._defaultAgent = globalAgent;
-  return new ClientRequest(options, cb);
+  args.unshift(options);
+
+  return new ClientRequest(...args);
```

Comment:
Last I checked, this was still significantly slower in V8.

Reply:
Fixed in Node 8.3?  See https://github.com/davidmarkclements/v8-perf

> Here’s the takeaway: if we want to write performant code around processing function inputs as an array (which in my experience seems fairly common), then in Node 8.3 and up we should use the spread operator.

Responding to your original suggestions:

>  Making HttpsClientRequest have no additional methods and checking this.constructor === HttpsClientRequest

The issue is that without additional methods, it never would be called... unless we duplicated the entire constructor.

> Allow passing a 4th option to ClientRequest that if not undefined, must be a private symbol indicating an https client request

This will impact the ability to add future arguments.  And we are currently dealing with three arguments, all optional...

 > Make the few changes needed to the original code in https to support both url and options

I went this way for now.  

The other option:

> Maybe we should move them under a private Symbol?

This is clean and would work.  The symbols would need to be exported by `lib/_http_client.js`

Reply:
@mcollina I'm not following.... in particular, I'm not sure what 'it' refers to.

Should I go back to the previous approach of subclassing ClientRequest for https, and introducing two new methods, this time with the method names being Symbols?  My thoughts are that that approach is more understandable and maintainable.

Or would you prefer the current approach which duplicates a bit of code, but doesn't introduce any new methods.  My thoughts are that approach is more private.

Reply:
I was actually commenting on some old code. The current approach is fine. LGTM

Reply:
> > Making HttpsClientRequest have no additional methods and checking this.constructor === HttpsClientRequest

> The issue is that without additional methods, it never would be called... unless we duplicated the entire constructor.

I'm not sure what you mean, `HttpsClientRequest` would extend from `ClientRequest` and be empty itself (no explicit constructor or additional methods). Then inside `ClientRequest`, checking `this.constructor` like:

```js
var defaultAgent = options._defaultAgent ||
                   (this.constructor === https[ClientRequestSym]
                    ? https.globalAgent : Agent.globalAgent);
```

Reply:
@mcdex Fair point, but making that work would involve a [circular require](https://nodejs.org/api/modules.html#modules_cycles).

Reply:
@rubys Not if it's lazy loaded.